cpp/src/neighbors/ann_mg/ann_mg.cuh

```diff
@@ -0,0 +1,747 @@
+            in_distances.data_handle() + batch_offset, n_rows_of_current_batch, n_neighbors);
+          ann_if.search(dev_res,
+                        search_params,
+                        query_partition,
+                        d_neighbors,
+                        d_distances);  // write search results inplace
+
+          // wait for results of other ranks
+          RAFT_NCCL_TRY(ncclGroupStart());
+          for (int from_rank = 0; from_rank < num_ranks_; from_rank++) {
+            if (from_rank == clique.root_rank_) continue;
+
+            batch_offset = from_rank * n_rows_of_current_batch * n_neighbors;
+            comms.device_recv(in_neighbors.data_handle() + batch_offset,
+                              n_rows_of_current_batch * n_neighbors,
+                              from_rank,
+                              resource::get_cuda_stream(dev_res));
+            comms.device_recv(in_distances.data_handle() + batch_offset,
+                              n_rows_of_current_batch * n_neighbors,
+                              from_rank,
+                              resource::get_cuda_stream(dev_res));
+          }
+          RAFT_NCCL_TRY(ncclGroupEnd());
+          resource::sync_stream(dev_res);
+        } else {  // non-root ranks
+          auto d_neighbors = raft::make_device_matrix<IdxT, IdxT, row_major>(
+            dev_res, n_rows_of_current_batch, n_neighbors);
+          auto d_distances = raft::make_device_matrix<float, IdxT, row_major>(
+            dev_res, n_rows_of_current_batch, n_neighbors);
+          ann_if.search(
+            dev_res, search_params, query_partition, d_neighbors.view(), d_distances.view());
+
+          // send results to root rank
+          RAFT_NCCL_TRY(ncclGroupStart());
+          comms.device_send(d_neighbors.data_handle(),
+                            n_rows_of_current_batch * n_neighbors,
+                            clique.root_rank_,
+                            resource::get_cuda_stream(dev_res));
+          comms.device_send(d_distances.data_handle(),
+                            n_rows_of_current_batch * n_neighbors,
+                            clique.root_rank_,
+                            resource::get_cuda_stream(dev_res));
+          RAFT_NCCL_TRY(ncclGroupEnd());
+          resource::sync_stream(dev_res);
+        }
+      }
+#pragma omp barrier
+
+      const auto& root_handle_ = clique.set_current_device_to_root_rank();
+      auto h_trans             = std::vector<IdxT>(num_ranks_);
+      IdxT translation_offset  = 0;
+      for (int rank = 0; rank < num_ranks_; rank++) {
+        h_trans[rank] = translation_offset;
+        translation_offset += ann_interfaces_[rank].size();
+      }
+      auto d_trans = raft::make_device_vector<IdxT, IdxT>(root_handle_, num_ranks_);
+      raft::copy(
+        d_trans.data_handle(), h_trans.data(), num_ranks_, resource::get_cuda_stream(root_handle_));
+
+      cuvs::neighbors::detail::knn_merge_parts(in_distances.data_handle(),
```

Comment:
Do I understand correctly, that all threads participate here, each merge a part of the data?

Reply:
Unless mistaken, this section of code is supposed to run on the main thread aside of OpenMP. The merge should happen on the GPU designated as "root".

Reply:
You are right. I have misread the code and did not see that the omp parallel region is already closed.